### PR TITLE
WIP: High-rate IMU sampling

### PR DESCRIPTION
**Pull request by AndreasAntener**
Continuing https://github.com/PX4/Firmware/pull/7446. @dagar @priseborough @jgoppert . I needed a better idea about vibration in a system and this is a good start. This is also one step towards https://github.com/PX4/Firmware/pull/7446#issuecomment-354678120

This still suffers from missing DMA/FIFO access. But I was able to log unfiltered data from 2 IMUs at 2kHz, which is already pretty nice. And you can publish gyro control messages at 2kHz as well, but in that case the MC attitude controller uses 22% CPU ;)

This is NOT test flown, I'm really just bench testing at the moment.

Tasks / open questions
- [x] really disable DLPF
- [x] separation done for MPU6000 and MPU9250
- [x] separate gyro message for control and estimation
- [x] add message to log unfiltered (but NOT raw) data for spectral analysis
- [ ] clip detector (should that be done in the driver?)
- [ ] do everything for the rest of gyro/accel drivers
- [ ] optimize message used for estimation: do we need RAW data there? (do we need to publish raw data at all?)
- [ ] add interface for new driver options (e.g. to enable unfiltered publishing)
- [ ] check ISR load, probably the cause for increased CPU load in every task when increasing the publishing rates
- [ ] read data with DMA/FIFO

Is that going into the right direction?
@dagar you mentioned you could help with DMA?

### CODE REVIEW

File: src/drivers/drv_accel.h [hunk 1]
```diff
@@ -52,7 +52,9 @@
 #define ACCEL2_DEVICE_PATH	"/dev/accel2"
 
 #include <uORB/topics/sensor_accel.h>
+#include <uORB/topics/sensor_accel_unfiltered.h>
 #define accel_report sensor_accel_s
+#define accel_unfiltered_report sensor_accel_unfiltered_s
```
Comment (dagar) [inline src/drivers/drv_accel.h]: I think we can stop doing this now.

File: src/drivers/mpu6000/mpu6000.cpp [hunk 2]
```diff
@@ -199,6 +201,13 @@ class MPU6000 : public device::CDev
 	float			_gyro_range_rad_s;
 
 	unsigned		_sample_rate;
+	unsigned		_gyro_control_interval;
+	unsigned		_gyro_count;
+	bool			_gyro_publish_unfiltered;
+	perf_counter_t		_sample_interval_accel;
```
Comment (dagar) [inline src/drivers/mpu6000/mpu6000.cpp]: I would consider consolidating these are dropping as many as possible. If you profile a nuttx board you'll find they definitely aren't free.

**Comment by dagar** [comment]
Questions
 - is logging raw feasible under normal operation? Do we need queuing for that uorb message or an or use a single message with an array of values

**Comment by dagar** [comment]
Rebased on master and added a trivial formatting fix.

**Comment by dagar** [comment]
If we only run the attitude control portion with attitude updates (https://github.com/PX4/Firmware/pull/8731/commits/8ca03021f59160a5b693fd7e9d104f02da6e189d), the mc_att_control cpu usage is 17% at 2kHz and ~8-9% at 1kHz.

What's interesting though is that the actual work of the rate controller (copy gyro data, PID, publish) is incredibly small (~ 25us). It's something else in the system (uORB, poll, etc) that's killing us.
https://github.com/PX4/Firmware/issues/6840

	sensor_baro                      0    3   68     0 1
	sensor_accel                     0    1  250     0 1
	sensor_gyro                      0    1  254     0 1
	sensor_gyro_control              0    1 1020     0 1
	sensor_gyro_unfiltered           0    0 2040     0 1
	sensor_gyro                      1    1  237     8 1
	sensor_mag                       0    1   99     0 1
	sensor_accel                     1    1  234     0 1
	adc_report                       0    1  100     0 1
	system_power                     0    2  100    11 1
	vehicle_control_mode             0    8    4     0 1
	actuator_controls_0              0    8 1020  6327 1
	sensor_combined                  0    6  254   928 1
	sensor_preflight                 0    1  254     0 1
	battery_status                   0    8   84   273 1
	vehicle_status                   0   10    4     0 1
	actuator_armed                   0    9    4     2 1
	safety                           0    1   49     0 1
	manual_control_setpoint          0    5    4     0 1
	vehicle_local_position           0    8  254  1716 1
	vehicle_attitude                 0    8  254  1040 1
	estimator_status                 0    5  254  2335 1
	commander_state                  0    0    4     0 1
	servorail_status                 0    0   49     0 1
	actuator_outputs                 0    4   49   201 1
	multirotor_motor_limits          0    1   49     0 1
	actuator_outputs                 1    2  347   321 1
	ekf2_timestamps                  0    1  254     0 1
	vehicle_rates_setpoint           0    2  254     0 1
	sensor_bias                      0    4  254   754 1
	task_stack_info                  0    0    1     0 2
	wind_estimate                    0    4  254   240 1
	ekf2_innovations                 0    1  254     0 1
	mc_att_ctrl_status               0    0 1020     0 1

**Comment by AndreasAntener** [comment]
@dagar I see you already added the mc att control update too, cool :). Yeah the load must be somewhere else, I was pretty surprised when I saw 22% for not doing so much..

>     is logging raw feasible under normal operation? Do we need queuing for that uorb message or an or use a single message with an array of values

It would definitely need some more magic to do this. Currently, for logging 2 IMUs at 2kHz I needed the logger to run at 4kHz, and even with a larger buffer I see periodic drops at the write operation (I think). I'm using the fastest possible SD cards and only logging the IMU data. The logger uses ~20% CPU in that case. At least for VTOL I don't see this happening at the moment with a v2/v3. But as we see there is a lot of cpu power vanishing, so lets have a look how this develops.

**Comment by dagar** [comment]
Closing this PR, but keeping the branch for now.
New effort (with SPI DMA) in #10602

**Pull request closed without merging**